### ISSUE TITLE
Query result cache invalidation

### ISSUE THREAD

**Issue by igorjos** [felixge/node-mysql#1175]
Hi I found this plugin very handy for my node.js app, but I have a problem with query invalidation.
Once I execute the query the result still stays stored like in cache... so no matter what I send next to that same query the result will be the same as the first time.

I'm trying not to close the connection, instead of that I want to use permanent connection, and execute the same select query but with different parameters, and to get different result.

In my current node configuration, the mysql connection is executed only once when node http server starts. Far I saw in the documentation says that connection.query will automatically execute connection.connect, but if after each connection.query I do connection.end then I'm receiving an error that says that connection haven't been re-executed again.

**Comment by igorjos** [comment]
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: Cannot enqueue Handshake after invoking quit.
    at Protocol._validateEnqueue (C:\xampp\htdocs\projects\nodesrv\node_modules\
mysql\lib\protocol\Protocol.js:196:16)
    at Protocol._enqueue (C:\xampp\htdocs\projects\nodesrv\node_modules\mysql\li
b\protocol\Protocol.js:129:13)
    at Protocol.handshake (C:\xampp\htdocs\projects\nodesrv\node_modules\mysql\l
ib\protocol\Protocol.js:52:41)
    at Connection.connect (C:\xampp\htdocs\projects\nodesrv\node_modules\mysql\l
ib\Connection.js:123:18)
    at Object.authorization.login (C:\xampp\htdocs\projects\nodesrv\libs\auth.js
:19:8)
    at C:\xampp\htdocs\projects\nodesrv\dispatcher.js:25:8
    at C:\xampp\htdocs\projects\nodesrv\node_modules\httpdispatcher\httpdispatch
er.js:130:3
    at HttpChain.next (C:\xampp\htdocs\projects\nodesrv\node_modules\httpdispatc
her\httpdispatcher.js:123:9)
    at doDispatch (C:\xampp\htdocs\projects\nodesrv\node_modules\httpdispatcher\
httpdispatcher.js:58:13)
    at IncomingMessage.<anonymous> (C:\xampp\htdocs\projects\nodesrv\node_module
s\httpdispatcher\httpdispatcher.js:69:15)

**Comment by dougwilson** [comment]
> then I'm receiving an error that says that connection haven't been re-executed again.

This is because a `connection` object can only ever be used one time: and you can `.end()` on it, it's useless and you have to `createConnection()` to create a new connection to the server.

**Comment by igorjos** [comment]
Ok, then what do you suggest to be used in order to invalidate previous result, without ending the connection?

**Comment by dougwilson** [comment]
I've never even heard of the behavior you're describing, so I'm not sure. Can you point me to the section in the MySQL manual that describes query result caching? Perhaps I can figure something out from there.

**Comment by dougwilson** [comment]
Otherwise, you can always reset the session state on a connection using the `.changeUser()` command (https://github.com/felixge/node-mysql#switching-users-and-altering-connection-state).

**Comment by igorjos** [comment]
I mentioned that it acts like it use cache, I'm not sure is there any cache involved.

This is my code:

define(['db', 'localStorage', 'MD5', 'jquery-deferred'], function (db, localStorage, md5, $) {
    
	"use strict";
	var deferred = $.Deferred();
	 
	var authorization = {
				
		login: function(request)
		{
			try
			{
				db.query('SELECT users.id,\
								 users.fid,\
								 users.name,\
								 users.email,\
								 users.banned,\
								 users.joined_at,\
								 NOW() as time\
							FROM users \
							WHERE email = '+db.escape(request.params.username)+'\
							AND password = '+db.escape(md5(request.params.password))+' ', function(err, rows, fields) {
				  
					  if (err) deferred.reject({'error': err});
					  
					  if(rows.length > 0)
					  {
						localStorage.setItem('user_'+rows[0].id, rows);
						deferred.resolve(rows);
					  }
					  else
					  {
						deferred.reject({'error': 'Nothing found'});
					  }
					  					  					  
				});
								
			}
			catch(error)
			{
				deferred.reject({'error': 'Login DB error'});
				return false;
			}
			return deferred.promise();
		},
		
	};
	
	return authorization;
});

When I execute auth.login for the first time, it is working great, but on the second time, it returns the same result as from the first time, even if I change all the parameters.
I make sure that correct parameters are sent, and did a lot of debugging, but the result is still the same. I will try your suggestion now and see if there are any changes.

Thanks

**Comment by igorjos** [comment]
Nope even changeUser doesn't help. It still returns the old result.

**Comment by dougwilson** [comment]
Yea, I'm not sure. The only way I can answer is I would need to have a way to reproduce the issue, if you can provide full complete code with the MySQL DDL to use and instructions on how to reproduce. Otherwise I'm not sure, especially since I've never hear about the behavior you describe.

You can also rule out that it's even this module by replacing your `mysql` dependency with the `mysql2` module, which uses a different protocol to interact with MySQL but has the same API.

Let me know when you can do either one of those :)

**Comment by igorjos** [comment]
/* DB file */

define(['mysql'], function (mysql) {
	
	var connection = mysql.createConnection({
	  host     : 'localhost',
	  user     : 'root',
	  password : '',
	  database : 'pickmeup',
	  
	});

	return connection;

});


Save this file as db.js or whatever, and use it in another file under node, then try to execute the same select query just with different values in the where clause. No matter is mysql or mysql2 the result will be the same... 

Instancing that file
define(['db'], function (db)
{
 db.query(.....);
}

**Comment by dougwilson** [comment]
> No matter is mysql or mysql2 the result will be the same... 

If this is the case, what would make this an issue with this module, if it also happens with a module that does not share any of the same code? I mean, if you tried `mysql2` and got the same results, it would indicate that the caching is either happening somewhere in your code or how you're using MySQL?

> Save this file as db.js or whatever, and use it in another file under node, then try to execute the same select query just with different values in the where clause

I'm not able to execute your SELECT statement without creating the necessary table first, can you provide the DDL and an INSERT so I have something to select?

Also, trying that code, I get the following error:
```
define(['mysql'], function (mysql) {
^
ReferenceError: define is not defined
    at Object.<anonymous> (db.js:1:1)
```

I've never heard of the function `define` in Node.js. Can you provide full, complete code with full, complete instructions? So far these are just pieces. I would like to help you, but you have to provide me a way to even try.

**Comment by igorjos** [comment]
I'm afraid that I can not place the whole project and its structure because it is in a late phase of development. The define[] method is used with requirejs.

Anyways I will try to separate the files and send you only needed files, so you can check on your local environment.

Thanks.

**Comment by igorjos** [comment]
Just as an info, I found a solution to the problem. Actually the deferred was the one that caused this issue. It was not waiting for the db.query to finish 2nd time. So instead of having $.Deferred declared globaly I have move it to be a part of the .login method. 

From my example above:

var deferred = $.Deferred();

should be now under
login:{
var deferred = $.Deferred();
.....
}

I hope that this will help the others who are using requirejs and jquery deferred together with mysql :)

**Comment by dougwilson** [comment]
Ah, I see! I saw you using it, but I've never used jQuery deferred, so had no idea if it was right or not. I have never done front-end work and am not generally familiar with jQuery (I have also never seen anyone using that on the backend before).
